File: tools/reference-implementations/python/vlp_secure.py
"""Secure-transport reference for VLP v0.2.

Implements:

* HKDF-SHA256 key derivation (stdlib `hashlib` + `hmac` — no third-party deps).
* ChaCha20-Poly1305 shared-key and master-key seal/open (requires the
  third-party ``cryptography`` package — install via ``pip install cryptography``).

Non-normative. See `book/src/spec/vlp-secure.md` for the authoritative
spec.
"""

from __future__ import annotations

import hashlib
import hmac
import struct


# ---------------------------------------------------------------------------
# HKDF-SHA256 (RFC 5869)
# ---------------------------------------------------------------------------


def hkdf_extract(salt: bytes, ikm: bytes) -> bytes:
    if not salt:
        salt = b"\x00" * hashlib.sha256().digest_size
    return hmac.new(salt, ikm, hashlib.sha256).digest()


def hkdf_expand(prk: bytes, info: bytes, length: int) -> bytes:
    hash_len = hashlib.sha256().digest_size
    if length > 255 * hash_len:
        raise ValueError("HKDF expand: requested length too large")
    out = b""
    t = b""
    counter = 1
    while len(out) < length:
        t = hmac.new(prk, t + info + bytes([counter]), hashlib.sha256).digest()
        out += t
        counter += 1
    return out[:length]


def hkdf_sha256(ikm: bytes, salt: bytes, info: bytes, length: int) -> bytes:
    return hkdf_expand(hkdf_extract(salt, ikm), info, length)


# ---------------------------------------------------------------------------
# Domain-specific key derivations (mirror book/src/spec/vlp-secure.md §6).
# ---------------------------------------------------------------------------


def derive_agent_key(master_key: bytes, agent_id: int) -> bytes:
    """HKDF-SHA256 derive_agent_key. Returns 32 bytes."""
    assert len(master_key) == 32
    info = b"varta-agent-v1\x00" + struct.pack("<I", agent_id)
    assert len(info) == 19, "info string length should be 15 + 4 = 19"
    return hkdf_sha256(ikm=master_key, salt=b"", info=info, length=32)


def derive_iv_prefix(session_salt: bytes, prefix_index: int) -> bytes:
    """HKDF-SHA256 derive_iv_prefix. Returns 8 bytes.

    The Rust reference passes ``None`` (empty salt) to HKDF-Extract; the
    16-byte session salt is the IKM. See `book/src/spec/vlp-secure.md` §6.2.
    """
    assert len(session_salt) == 16
    info = b"varta-iv-prefix-v1\x00" + struct.pack("<I", prefix_index)
    assert len(info) == 23
    return hkdf_sha256(ikm=session_salt, salt=b"", info=info, length=8)


def derive_epoch_key(agent_key: bytes, epoch: int) -> bytes:
    """HKDF-SHA256 derive_epoch_key. Returns 32 bytes."""
    assert len(agent_key) == 32
    info = b"varta-epoch-v1\x00" + struct.pack("<Q", epoch)
    assert len(info) == 23
    return hkdf_sha256(ikm=agent_key, salt=b"", info=info, length=32)


# ---------------------------------------------------------------------------
# AEAD wrapping (requires ``cryptography``).
# ---------------------------------------------------------------------------


def _aead():
    """Lazy import so the base-frame Python ref does not require pip deps."""
    from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
    return ChaCha20Poly1305


def encode_shared(key: bytes, iv_random: bytes, iv_counter: int, plaintext: bytes) -> bytes:
    """Produce a 60-byte shared-key secure frame."""
    assert len(key) == 32 and len(iv_random) == 8 and len(plaintext) == 32
    nonce = iv_random + struct.pack("<I", iv_counter)
    ct_and_tag = _aead()(key).encrypt(nonce, plaintext, associated_data=None)
    return iv_random + struct.pack("<I", iv_counter) + ct_and_tag


def decode_shared(key: bytes, wire: bytes) -> bytes:
    """Recover the 32-byte plaintext frame from a 60-byte shared-key wire frame."""
    assert len(wire) == 60
    iv_random = wire[0:8]
    iv_counter = struct.unpack("<I", wire[8:12])[0]
    ct_and_tag = wire[12:60]
    nonce = iv_random + struct.pack("<I", iv_counter)
    return _aead()(key).decrypt(nonce, ct_and_tag, associated_data=None)


def encode_master(master_key: bytes, agent_pid: int, iv_random: bytes,
                  iv_counter: int, plaintext: bytes) -> bytes:
    """Produce a 64-byte master-key secure frame."""
    assert len(master_key) == 32 and len(iv_random) == 8 and len(plaintext) == 32
    agent_key = derive_agent_key(master_key, agent_pid)
    aad = struct.pack("<I", agent_pid)
    nonce = iv_random + struct.pack("<I", iv_counter)
    ct_and_tag = _aead()(agent_key).encrypt(nonce, plaintext, associated_data=aad)
    return aad + iv_random + struct.pack("<I", iv_counter) + ct_and_tag


def decode_master(master_key: bytes, wire: bytes) -> bytes:
    """Recover the 32-byte plaintext from a 64-byte master-key wire frame."""
    assert len(wire) == 64
    aad = wire[0:4]
    agent_pid = struct.unpack("<I", aad)[0]
    iv_random = wire[4:12]
    iv_counter = struct.unpack("<I", wire[12:16])[0]
    ct_and_tag = wire[16:64]
    agent_key = derive_agent_key(master_key, agent_pid)
    nonce = iv_random + struct.pack("<I", iv_counter)
    return _aead()(agent_key).decrypt(nonce, ct_and_tag, associated_data=aad)

File: tools/reference-implementations/go/vlp_secure.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"fmt"

	"golang.org/x/crypto/chacha20poly1305"
)

// ----------------------------------------------------------------------------
// HKDF-SHA256 (RFC 5869) — small enough to inline; uses stdlib hash + hmac.
// ----------------------------------------------------------------------------

func hkdfExtract(salt, ikm []byte) []byte {
	if len(salt) == 0 {
		salt = make([]byte, sha256.Size)
	}
	mac := hmac.New(sha256.New, salt)
	mac.Write(ikm)
	return mac.Sum(nil)
}

func hkdfExpand(prk, info []byte, length int) []byte {
	if length > 255*sha256.Size {
		panic(fmt.Sprintf("hkdf: requested length %d too large", length))
	}
	var out, t []byte
	counter := byte(1)
	for len(out) < length {
		mac := hmac.New(sha256.New, prk)
		mac.Write(t)
		mac.Write(info)
		mac.Write([]byte{counter})
		t = mac.Sum(nil)
		out = append(out, t...)
		counter++
	}
	return out[:length]
}

func hkdfSha256(ikm, salt, info []byte, length int) []byte {
	return hkdfExpand(hkdfExtract(salt, ikm), info, length)
}

// ----------------------------------------------------------------------------
// Domain-specific derivations (mirror book/src/spec/vlp-secure.md §6).
// ----------------------------------------------------------------------------

// DeriveAgentKey returns the 32-byte per-agent key.
func DeriveAgentKey(masterKey [32]byte, agentID uint32) [32]byte {
	info := make([]byte, 19)
	copy(info[:15], []byte("varta-agent-v1\x00"))
	binary.LittleEndian.PutUint32(info[15:], agentID)
	okm := hkdfSha256(masterKey[:], nil, info, 32)
	var out [32]byte
	copy(out[:], okm)
	return out
}

// DeriveIVPrefix returns the 8-byte per-session IV prefix.
func DeriveIVPrefix(sessionSalt [16]byte, prefixIndex uint32) [8]byte {
	info := make([]byte, 23)
	copy(info[:19], []byte("varta-iv-prefix-v1\x00"))
	binary.LittleEndian.PutUint32(info[19:], prefixIndex)
	okm := hkdfSha256(sessionSalt[:], nil, info, 8)
	var out [8]byte
	copy(out[:], okm)
	return out
}

// DeriveEpochKey returns the 32-byte per-epoch key.
func DeriveEpochKey(agentKey [32]byte, epoch uint64) [32]byte {
	info := make([]byte, 23)
	copy(info[:15], []byte("varta-epoch-v1\x00"))
	binary.LittleEndian.PutUint64(info[15:], epoch)
	okm := hkdfSha256(agentKey[:], nil, info, 32)
	var out [32]byte
	copy(out[:], okm)
	return out
}

// ----------------------------------------------------------------------------
// AEAD wrapping (60-byte shared-key / 64-byte master-key).
// ----------------------------------------------------------------------------

// EncodeShared produces a 60-byte shared-key secure frame.
func EncodeShared(key [32]byte, ivRandom [8]byte, ivCounter uint32, plaintext [32]byte) ([60]byte, error) {
	cipher, err := chacha20poly1305.New(key[:])
	if err != nil {
		var z [60]byte
		return z, err
	}
	nonce := make([]byte, 12)
	copy(nonce[:8], ivRandom[:])
	binary.LittleEndian.PutUint32(nonce[8:], ivCounter)
	ctAndTag := cipher.Seal(nil, nonce, plaintext[:], nil)

	var out [60]byte
	copy(out[0:8], ivRandom[:])
	binary.LittleEndian.PutUint32(out[8:12], ivCounter)
	copy(out[12:60], ctAndTag)
	return out, nil
}

// EncodeMaster produces a 64-byte master-key secure frame.
func EncodeMaster(masterKey [32]byte, agentPID uint32, ivRandom [8]byte, ivCounter uint32, plaintext [32]byte) ([64]byte, [32]byte, error) {
	agentKey := DeriveAgentKey(masterKey, agentPID)
	cipher, err := chacha20poly1305.New(agentKey[:])
	if err != nil {
		var z [64]byte
		return z, agentKey, err
	}
	aad := make([]byte, 4)
	binary.LittleEndian.PutUint32(aad, agentPID)
	nonce := make([]byte, 12)
	copy(nonce[:8], ivRandom[:])
	binary.LittleEndian.PutUint32(nonce[8:], ivCounter)
	ctAndTag := cipher.Seal(nil, nonce, plaintext[:], aad)

	var out [64]byte
	copy(out[0:4], aad)
	copy(out[4:12], ivRandom[:])
	binary.LittleEndian.PutUint32(out[12:16], ivCounter)
	copy(out[16:64], ctAndTag)
	return out, agentKey, nil
}
